game/graphics/texture/TextureConverter.cpp

```diff
@@ -45,167 +41,67 @@ void TextureConverter::download_rgba8888(u8* result,
-    // not sure what the other widths are yet.
-    int read_width = 64 * goal_tex_width;
-
-    // loop over pixels in output texture image
-    for (u32 y = 0; y < h; y++) {
-      for (u32 x = 0; x < w; x++) {
-        // read as the PSMT8 type. The dest field tells us a block offset.
-        auto addr8 = psmt8_addr(x, y, read_width) + vram_addr * 256;
-        u8 value = *(u8*)(m_vram.data() + addr8);
-
-        // there's yet another scramble from the CLUT. The palette index turns into an X, Y value
-        // See GS manual 2.7.3 CLUT Storage Mode, IDTEX8 in CSM1 mode.
-        u32 clut_chunk = value / 16;
-        u32 off_in_chunk = value % 16;
-        u8 clx = 0, cly = 0;
-        if (clut_chunk & 1) {
-          clx = 8;
-        }
-        cly = (clut_chunk >> 1) * 2;
-        if (off_in_chunk >= 8) {
-          off_in_chunk -= 8;
-          cly++;
-        }
-        clx += off_in_chunk;
-
-        // the x, y CLUT value is looked up in PSMCT32 mode
-        u32 clut_addr = psmct16_addr(clx, cly, 64) + clut_vram_addr * 256;
-        u32 clut_value = *(u16*)(m_vram.data() + clut_addr);
-        u32 rgba32 = rgba16_to_rgba32(clut_value);
-        memcpy(result + out_offset, &rgba32, 4);
-        out_offset += 4;
-      }
+    clx += off_in_chunk;
+    if (clut_psm == int(CPSM::PSMCT32)) {
+      return psmct32_addr(clx, cly, 64) + clut_vram_addr * 256;
+    } else if (clut_psm == int(CPSM::PSMCT16)) {
+      return psmct16_addr(clx, cly, 64) + clut_vram_addr * 256;
     }
-
-  } else if (psm == int(PSM::PSMT4) && clut_psm == int(CPSM::PSMCT16)) {
-    // width is like the TEX0 register, in 64 texel units.
-    // not sure what the other widths are yet.
-    int read_width = 64 * goal_tex_width;
-
-    // loop over pixels in output texture image
-    for (u32 y = 0; y < h; y++) {
-      for (u32 x = 0; x < w; x++) {
-        // read as the PSMT4 type, use half byte addressing
-        auto addr4 = psmt4_addr_half_byte(x, y, read_width) + vram_addr * 512;
-
-        // read (half bytes)
-        u8 value = *(u8*)(m_vram.data() + addr4 / 2);
-        if (addr4 & 1) {
-          value >>= 4;
-        } else {
-          value = value & 0x0f;
-        }
-
-        // there's yet another scramble from the CLUT. The palette index turns into an X, Y value
-        // See GS manual 2.7.3 CLUT Storage Mode, IDTEX4 in CSM1 mode.
-
-        u8 clx = value & 0x7;
-        u8 cly = value >> 3;
-
-        // the x, y CLUT value is looked up in PSMCT16 mode
-        u32 clut_addr = psmct16_addr(clx, cly, 64) + clut_vram_addr * 256;
-        u32 clut_value = *(u16*)(m_vram.data() + clut_addr);
-        u32 rgba32 = rgba16_to_rgba32(clut_value);
-        memcpy(result + out_offset, &rgba32, 4);
-        out_offset += 4;
-      }
+    ASSERT(false);  // Invalid CLUT format
+    return 0;
+  };
+
+  // Pixel processing function
+  auto process_pixel = [&](u32 x, u32 y, u32 addr, int clut_psm) -> u32 {
```

Comment:
This comment is not very meaningful, and it's not intuitive to figure out what this function does. Is `addr` an address into the clut or the index texture? What does it return? Something like
```
auto lookup_rgba_from_clut = [&](u32 index_texture_value, int clut_psm) -> u32
```
may be more clear.

Or, since this lambda is only used one, you could avoid creating this entirely.

Reply:
The lambda has been renamed, but since its called twice i avoided removing it